volumes/short_term_counting_program/move_match/speedvol_stats.sql

```diff
@@ -0,0 +1,251 @@
+        AND countinfo.count_date < '2023-01-15' -- short time period for testing purposes
+), 
+
+-- Calculate volume percentages and cumulative percentages by speed bin
+spd_bin_vol AS (
+    SELECT
+        spd_vol.arterycode,
+        spd_vol.count_info_id,
+        spd_vol.direction,
+        spd_vol.location,
+        spd_vol.datetime_bin::date AS dt,
+        spd_vol.speed_kph,
+        SUM(spd_vol.volume_15min) AS spd_vol, -- aggregate 15 minute bins into a daily count
+        SUM(SUM(spd_vol.volume_15min)) OVER cumulative / (SUM(SUM(spd_vol.volume_15min)) OVER _all + 0.00000001) AS cum_pct, -- SUM(SUM()) explained below
+        SUM(spd_vol.volume_15min) / (SUM(SUM(spd_vol.volume_15min)) OVER _all + 0.00000001) AS pct -- SUM(SUM()) explained below
+    FROM spd_vol
+    GROUP BY 
+        spd_vol.arterycode,
+        spd_vol.count_info_id,
+        spd_vol.direction,
+        spd_vol.location,
+        spd_vol.datetime_bin::date,
+        spd_vol.speed_kph
+    WINDOW
+        cumulative AS (PARTITION BY spd_vol.count_info_id ORDER BY spd_vol.speed_kph RANGE UNBOUNDED PRECEDING),
+        _all AS (PARTITION BY spd_vol.count_info_id)
+),
+
+/*
+SUM(SUM()) Explained
+There are two lines in the CTE above that use a nested aggregate function within a WINDOW function.
+A WINDOW function allows calculations to be performed on subsets of data (or WINDOWs of data) without collapsing the data into aggregated forms (like GROUP BY).
+In this CTE, two WINDOWs are used: 
+    cumulative: creates windows based on count_info_id and speed bin
+    _all: creates windows based on count_info_id
+Here is an explanation of the first SUM(SUM()) line:
+    Part 1 - SUM(SUM(spd_vol.volume_15min)) OVER cumulative: 
+        The inside SUM() adds up 15 minute volume data (resulting in daily volumes by speed bin).
+        The outside SUM() adds up the bins over the cumulative window.
+        In plain language: sum daily totals by speed bin, then calculate a cumulative sum for ascending speed bins.
+    Part 2 - / (SUM(SUM(spd_vol.volume_15min)) OVER _all + 0.00000001):
+        The inside SUM() adds up 15 minute volume data (resulting daily volumes by speed bin).
+        The outside SUM() adds up all of the speed bin volumes (resulting in daily volumes).
+        In plain language: calculate daily volumes.
+    Since Part 1 is divided by Part 2, you end up with cumulative proportions of daily volume by speed bin.
+Here is an explanation of the second SUM(SUM()) line:
+    SUM(spd_vol.volume_15min) / (SUM(SUM(spd_vol.volume_15min)) OVER _all + 0.00000001) AS pct:
+    Calculate daily volumes by speed bin and divide it by...
+    Daily total volumes (15 minute volume data are aggregated to speed bins in the inside sum, then they are totalled in the outside sum).
+    You end up with the proportion of daily volume in each speed bin.
+*/
+
+-- Calculate 15th percentile speed
+per_15 AS (
+    SELECT DISTINCT ON (spd_bin_vol.count_info_id)
```

Comment:
You need to pair an ORDER BY with DISTINCT ON to make sure this always returns the correct results even if the source data is not sorted. See: https://www.postgresql.org/docs/current/sql-select.html

> The DISTINCT ON expressions are interpreted using the same rules as for ORDER BY (see above). Note that the “first row” of each set is unpredictable unless ORDER BY is used to ensure that the desired row appears first.

Reply:
Resolved using
```
ORDER BY
    spd_bin_vol.count_info_id,
    spd_bin_vol.cum_pct
```